File: Design-Patterns-Using-Cpp/Facad/Facad.cpp
#include <iostream>

// Subsystem1
class Subsystem1 {
public:
    void operation1() {
        std::cout << "Subsystem1 operation1" << std::endl;
    }

    void operation2() {
        std::cout << "Subsystem1 operation2" << std::endl;
    }
};

// Subsystem2
class Subsystem2 {
public:
    void operation1() {
        std::cout << "Subsystem2 operation1" << std::endl;
    }

    void operation2() {
        std::cout << "Subsystem2 operation2" << std::endl;
    }
};

// SystemManagerFacade
class SystemManagerFacade {
private:
    Subsystem1 subsystem1;
    Subsystem2 subsystem2;

public:
    void operation() {
        subsystem1.operation1();
        subsystem1.operation2();
        subsystem2.operation1();
        subsystem2.operation2();
    }
};

// Client code
int main() {
    SystemManagerFacade facade;
    facade.operation();

    return 0;
}

/*
Subsystem1 operation1
Subsystem1 operation2
Subsystem2 operation1
Subsystem2 operation2
*/

File: Design-Patterns-Using-JavaScript/Facad/Facad.js
// Subsystem1
class Subsystem1 {
    operation1() {
        console.log("Subsystem1 operation1");
    }

    operation2() {
        console.log("Subsystem1 operation2");
    }
}

// Subsystem2
class Subsystem2 {
    operation1() {
        console.log("Subsystem2 operation1");
    }

    operation2() {
        console.log("Subsystem2 operation2");
    }
}

// SystemManagerFacade
class SystemManagerFacade {
    constructor() {
        this.subsystem1 = new Subsystem1();
        this.subsystem2 = new Subsystem2();
    }

    operation() {
        this.subsystem1.operation1();
        this.subsystem1.operation2();
        this.subsystem2.operation1();
        this.subsystem2.operation2();
    }
}

// Client code
const facade = new SystemManagerFacade();
facade.operation();

/*
Subsystem1 operation1
Subsystem1 operation2
Subsystem2 operation1
Subsystem2 operation2
*/
